api/types.proto

```diff
@@ -281,15 +281,40 @@ message UpdateConfig {
 	enum FailureAction {
 		PAUSE = 0;
 		CONTINUE = 1;
-		// TODO(aaronl): Add ROLLBACK as a supported failure mode.
-		// (#486)
+		ROLLBACK = 2;
 	}
 
 	// FailureAction is the action to take when an update failures.
-	// Currently, a failure is defined as a single updated task failing to
-	// reach the RUNNING state. In the future, there will be configuration
-	// to define what is treated as a failure (see #486 for a proposal).
 	FailureAction failure_action = 3;
+
+	// MonitoringPeriod indicates how long to monitor a task for failure
+	// after it is created. If the task fails by ending up in one of the
+	// states REJECTED, COMPLETED, or FAILED, within MonitoringPeriod
+	// from its creation, this counts as a failure. If it fails after
+	// MonitoringPeriod, it does not count as a failure. If
+	// MonitoringPeriod is unspecified, a default value will be used.
+	Duration monitoring_period = 4;
+
+	// AllowedFailureFraction is the fraction of tasks that may fail during
+	// an update before the failure action is invoked. Any task created by
+	// the current update which ends up in one of the states REJECTED,
+	// COMPLETED or FAILED within MonitoringPeriod from its creation counts
+	// as a failure. The number of failures is divided by the number of
+	// tasks being updated, and if this fraction is greater than
+	// AllowedFailureFraction, the failure action is invoked.
+	//
+	// If the failure action is CONTINUE, there is no effect.
+	// If the failure action is PAUSE, no more tasks will be updated until
+	// another update is started.
+	// If the failure action is ROLLBACK, the orchestrator will attempt to
+	// roll back to the previous service spec. If the AllowedFailureFraction
+	// threshold is hit during the rollback, the rollback will pause.
+	//
+	// TODO(aaronl): Should there be a separate failure threshold for
+	// rollbacks? Should there be a failure action for rollbacks (to allow
+	// them to do something other than pause when the rollback encounters
+	// errors)?
+	float allowed_failure_fraction = 5;
```

Comment:
I kind of like fractions but I can't see a particularly strong argument for one over the other.

Pausing on the first failure is very easy to do with either a fraction or an absolute threshold (`tolerated_failures` = 0 or `allowed_failure_fraction` = 0.0).